webui/react/src/stores/projects.tsx

```diff
@@ -1,125 +1,127 @@
 import { Map } from 'immutable';
-import React, { createContext, PropsWithChildren, useCallback, useContext, useState } from 'react';
+import React, {
+  createContext,
+  PropsWithChildren,
+  useCallback,
+  useContext,
+  useEffect,
+  useState,
+} from 'react';
 
 import { getWorkspaceProjects } from 'services/api';
 import { Project } from 'types';
 import handleError from 'utils/error';
 import { Loadable, Loaded, NotLoaded } from 'utils/loadable';
+import { observable, useObservable, WritableObservable } from 'utils/observable';
+
+class ProjectService {
+  private _projects: WritableObservable<Map<number, Project>> = observable(Map());
+  private _projectsByIndex: WritableObservable<Map<string, Project[]>> = observable(Map());
+
+  genWorkspaceKey = (workspaceId: number): string => `byworkspace-${workspaceId}`;
+
+  fetchWorkspaceProjects = async (
+    workspaceId: number,
+    canceler: AbortController,
+    forceFetch = false,
+  ): Promise<void> => {
+    if (!forceFetch && this._projectsByIndex.get().get(this.genWorkspaceKey(workspaceId))) return;
+    try {
+      const response = await getWorkspaceProjects(
+        {
+          id: workspaceId,
+          limit: 0,
+        },
+        { signal: canceler.signal },
+      );
+      // Prevent unecessary re-renders
+      if (!forceFetch && this._projectsByIndex.get().get(this.genWorkspaceKey(workspaceId))) return;
+      this._projectsByIndex.update((prevState: Map<string, Project[]>) => {
+        return prevState.set(this.genWorkspaceKey(workspaceId), response.projects);
+      });
+      this._projects.update((prevState: Map<number, Project>) => {
+        return prevState.withMutations((state: Map<number, Project>) => {
+          response.projects.forEach((proj) => state.set(proj.id, proj));
+        });
+      });
+    } catch (e) {
+      handleError(e);
+    }
+  };
 
-type UpdateProjectsByIndex = (fn: (ws: Map<string, Project[]>) => Map<string, Project[]>) => void;
+  getWorkspaceProject = (workspaceId: number): Project[] | undefined => {
```

Comment:
Better to return a `Loadable` at this level since you're kind of exposing the implementation here.